### PR TITLE
add contrib.ScalaJSWebpackModule that replicates sbt's scalajs-bundler (#410)

### PR DESCRIPTION
**Pull request by crimzie**
This PR completes issue #410 .

### CODE REVIEW

**Comment by lihaoyi** [comment]
Anyone want to help me review this? As long as someone takes a look I'm happy for it to go in

File: contrib/webpacklib/src/mill/contrib/ScalaJSWebpackModule.scala [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+              case z if z.getName.endsWith(".js") &&
+                !z.getName.startsWith("scala/") =>
+                JsDeps(Nil, Nil, Map(z.getName -> read(stream).toString))
+            }
+            .to[Seq]
+        stream.close()
+        deps
+      }
+      .toSeq
+      .foldLeft(JsDeps(Nil, Nil, Map.empty)) { _ ++ _ }
```
Comment (DavidGregory084) [inline contrib/webpacklib/src/mill/contrib/ScalaJSWebpackModule.scala]: I feel like this mega-expression could be broken down a bit

**Comment by DavidGregory084** [comment]
Looks good to me, just one big method calling chain I would break down a bit

(caveat: I've never used Scala.js for anything serious)

File: contrib/webpacklib/src/mill/contrib/ScalaJSWebpackModule.scala [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+  }
+
+  def webpack: Task[(ops.Path, ops.Path, Boolean) => Unit] = T.task {
+    (src: ops.Path, dst: ops.Path, opt: Boolean) =>
+      val outjs = dst / src.segments.last
+      val deps = sbtBundlerDeps()
+      val cfg = "webpack.config.js"
+      ops.cp(src, outjs)
+      writeBundleSources().apply(deps, dst)
+      writeWpConfig().apply(dst, cfg, outjs.toString, "out-bundle.js", opt)
```
Comment (raquo) [inline contrib/webpacklib/src/mill/contrib/ScalaJSWebpackModule.scala]: My wishlist:
(1) configurable bundle name and location
(2) ability to provide a path to custom webpack config file

Regarding (2), I think Mill should not be configuring webpack directly. I think letting users configure webpack via plain webpack config files is much more straightforward than via an indirect, limited DSL. Mill's role should be simply to pass any necessary information such as the paths of Scala.js output files to the config.

Comment (gavares) [reply contrib/webpacklib/src/mill/contrib/ScalaJSWebpackModule.scala]: Not having a replacement for scaljs-bundler is just about the only thing keeping me from using mill currently. Having this change merged would be fantastic.

That said, I agree with @raquo that users should at least have the option to configure webpack directly. It's fairly complicated and the plugin ecosystem for webpack is so diverse that a cusom DSL will never keep pace or satisfy all the needs that users will have.

@crimzie , as I said I'm really looking forward to having this committed. Let me know if you could use any help.

Comment (crimzie) [reply contrib/webpacklib/src/mill/contrib/ScalaJSWebpackModule.scala]: Hi, @raquo @gavares ! 

I was pretty busy lately and couldn't respond any earlier .

As I wrote in original issue #410, there is a problem of existing libs pulling in dependencies on js sources via sbt-scalajs-bundler. This here is a minimalistic solution for that sole problem. It wasn't originally intended as a full-fledged tool that would let users add their own js dependencies like sbt-scalajs-bundler does and as such doesn't provide customization options.

Yet I agree with your arguments and will rewrite it with an option for user to provide config, just need to find some time to do it :)

**Comment by lihaoyi** [comment]
closing due to inactivity

**Comment by lefou** [comment]
@crimzie Any plan to revieve this one?

**Comment by nafg** [comment]
Updated version that I've used a little: https://gist.github.com/nafg/6ecce298a0a20f1e4a259cdae5634060

**Comment by sequencer** [comment]
@nafg is possible for you to upstream this?

**Comment by nafg** [comment]
I'll try to post my version of the code soon

**Comment by sequencer** [comment]
Thank you!

**Comment by nafg** [comment]
Sorry for the delay. Here is what I have currently. I believe it works but mill is not currently my main build tool so it hasn't been used for production purposes.

https://gist.github.com/nafg/edb39f43daf2cd7ca8300b4252ace709

**Pull request closed without merging**